packages/openapi-spec-builder/src/openapi-spec-builder.ts

```diff
@@ -170,7 +170,34 @@ export class OperationSpecBuilder extends BuilderBase<OperationObject> {
    * @param name The name of the controller method implementing this operation.
    */
   withOperationName(name: string): this {
-    return this.withExtension('x-operation-name', name);
+    this.withExtension('x-operation-name', name);
+    this.setupOperationId();
+    return this;
+  }
+
+  /**
+   * Define the controller name (controller name).
+   *
+   * @param name The name of the controller containing this operation.
+   */
+  withControllerName(name: string): this {
+    this.withExtension('x-controller-name', name);
+    this.setupOperationId();
+    return this;
+  }
+
+  private setupOperationId() {
+    if (this._spec.operationId) return;
+    const controllerName = this._spec['x-controller-name'];
+    const operationName = this._spec['x-operation-name'];
+    if (controllerName && operationName) {
+      this._spec.operationId = controllerName + '$' + operationName;
```

Comment:
This will produce IDs like `Product$create`. Have you considered using `.` separator to produce `Product.create` instead? 

In LoopBack 3, we use the dot character, see
https://github.com/strongloop/loopback-swagger/blob/52ba102c3e5dcd653c6d0a3243b16f4a51ecc169/test/specgen/route-helper.test.js#L501-L504
```js
  it('route operationId DOES include model name.', function() {
    var doc = createAPIDoc({method: 'User.login'});
    expect(doc.operation.operationId).to.equal('User.login');
  });
```

Also what's the impact on client generators like https://github.com/swagger-api/swagger-codegen, is there any difference in how they treat both `$` and `.` ?

Reply:
Swagger clients such as https://github.com/swagger-api/swagger-js uses `operationId` as method names. Using `PingController.ping` will be confusing.

Reply:
From what I remember from the times where we were looking into swagger-codegen, they were normalizing operation ids and replacing characters like `.` to `_`. As a result, you would end up with a method like `PingController_ping`.

AFAICT from reading swagger-js code, it's doing the same.

https://github.com/swagger-api/swagger-js/blob/0cdc6488372547f69ca7372cc9ad44cf4bf92de5/src/helpers.js#L29-L38
```js
export function opId(operation, pathName, method = '', {v2OperationIdCompatibilityMode} = {}) {
  if (!operation || typeof operation !== 'object') {
    return null
  }
  const idWithoutWhitespace = (operation.operationId || '').replace(/\s/g, '')
  if (idWithoutWhitespace.length) {
    return escapeString(operation.operationId)
  }
  return idFromPathMethod(pathName, method, {v2OperationIdCompatibilityMode})
}
```

https://github.com/swagger-api/swagger-js/blob/0cdc6488372547f69ca7372cc9ad44cf4bf92de5/src/helpers.js#L5-L7

```js
const escapeString = (str) => {
  return str.replace(/[^\w]/gi, '_')
}
```

Could you please verify?

If my understanding is correct, then both `PingController$ping` and `PingController.ping` are going to produce the same method name `PingController_ping`.

The reason why I am going to such length here is that changing the format of generated operation ids would be a breaking change for our consumers.

Reply:
Good finding. I'll use `.` instead of `$` then.